pysnark/runtime.py

```diff
@@ -204,97 +204,203 @@ def __ge__(self, other): return (self-other).check_positive()
-        
-        if not is_base_value(divisor): return NotImplemented
-        
-        if divisor==0: raise ValueError("division by zero")
-        
-        quo = PrivVal(self.value//divisor)
-        rem = PrivVal(self.value-quo.value*divisor)
- 
-        rem.assert_positive(divisor.bit_length())
-        if divisor&(divisor-1)!=0: rem.assert_lt(divisor) # not needed for powers of two
-        quo.assert_positive()
-        (self-divisor*quo-rem).assert_zero()
-        
-        return (quo,rem)
+        """
+        Divides a LinComb with an integer or another LinComb and returns the quotient and the remainder
+        Costs 0 constraints to divide with an integer
+        Costs 3 constraints to divide with a LinComb
+        """
+        if isinstance(divisor, int):
+            if divisor == 0:
+                raise ValueError("Division by zero")
+            quo = ConstVal(self.value // divisor)
+            rem = ConstVal(self.value % divisor)
+            return (quo, rem)
+
+        if isinstance(divisor, LinComb):
+            if divisor.value == 0:
+                raise ValueError("Division by zero")
+            quo = PrivVal(self.value // divisor.value)
+            res = quo * divisor
+            rem = PrivVal(self.value - res.value)
+
+            # TODO: Prevent field overflow by a malicious prover
+            # Add a constraint ensuring quo * divisor + rem < backend.get_modulus()
+            # The check above works over the integers, but not in a finite field
+
+            add_constraint(quo, divisor, self - rem)
+            rem.assert_lt(divisor)
+            quo.assert_positive()
+            return (quo,rem)
+
+        return NotImplemented
     
     def __pow__(self, other, mod=None):
-        """ Exponentiation with public integral power p>=0 """
-        if mod!=None: raise ValueError("cannot provide modulus")
-        if not is_base_value(other): return NotImplemented
-        if other<0: raise ValueError("exponent cannot be negative", other)
-        if other==0: return LinComb.ONE
-        if other==1: return self
-        return self*pow(self, other-1)
+        """
+        Raises a LinComb to the power of an integer or a LinComb
+        Costs 0 constraints to raise to the power of an integer
+        Costs ? constraints to raise to the power of a LinComb
+        """
+        if mod != None:
+            raise ValueError("Cannot provide modulus")
+        if isinstance(other, int):
+            if other < 0:
+                raise ValueError("Exponent cannot be negative")
+            if other == 0:
+                return LinComb.ONE
+            if other == 1:
+                return self
+            return self * self ** (other - 1)
+
+        if isinstance(other, LinComb):
+            # Obliviously apply repeated squaring
+            from .branching import if_then_else
+
+            # Limit exponent to prevent Python crashing with bus error
+            if other.value.bit_length() > 5:
```

Comment:
Would be nice if this would be somehow configurable

Reply:
What would be a good way to handle it? Another global variable? An extra parameter?

What if we update each multiplicand with `multiplicand.value = multiplicand.value % backend.get_modulus()`? Then, the maximum exponent would be limited only by `bitlength`. That would make the output LinComb's `value` and `__repr__` wrong, but would increase performance and shouldn't affect the actual circuit. 

It might also be worth thinking about how to handle `LinComb.value` when it overflows. Should we leave the value as is or wrap around? If we were to wrap the value around (e.g. in `__init__`), it might easily solve these issues.

Reply:
Yes, doing multiplicand.value = multiplicand.value % backend.get_modulus() definitely makes sense. Of course other.to_bits still requires a bit length, but I guess we can just use the default bit length and if the user wants to have a lower bit length (because of efficiency) he can just temporarily change it.

Reply:
And indeed, it is a good point about the overflows. Originally the code used to just always reduce modulo backend.get_modulus(). But later my idea was to keep stuff completely independent from the modulus wherever possible. Because who knows, one day there may be backend that just works over the integers and doesn't use a modulus at all. In fact, the only place where get_modulus is now used is in the hash code. 

Maybe instead of using backend.get_modulus(), the backends should instead support a backend.reduce(val) function that performs modular reduction...

Reply:
Should I just reduce using `backend.get_modulus()` for now?

Reply:
Yes, let's just keep it like that at least for __pow__. It doesn't make much sense to do a __pow__ that has an overflow, so in practice it shouldn't matter.

Reply:
I made the change, but oblivious pow now costs 7 * bitlength + 1 constraints. Is that acceptable?